torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,220 @@
+r"""
+This class defines the abstraction for a Graph feature store. The goal of a
+feature store is to abstract away all node and edge feature memory management
+so that varying implementations can allow for independent scale-out.
+
+This particular feature store abstraction makes a few key assumptions:
+    * The features we care about storing are all associated with some sort of
+        `index`; explicitly for PyG the the index of the node in the graph (or
+            the heterogeneous component of the graph it resides in).
+    * A feature can uniquely be identified from (a) its index and (b) any other
+        associated attributes specified in :obj:`TensorAttr`.
+
+It is the job of a feature store implementor class to handle these assumptions
+properly. For example, a simple in-memory feature store implementation may
+concatenate all metadata values with a feature index and use this as a unique
+index in a KV store. More complicated implementations may choose to partition
+features in interesting manners based on the provided metadata.
+
+Major TODOs for future implementation:
+* Async `put` and `get` functionality
+"""
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from typing import Any, Optional
+
+import numpy as np
+import torch
+
+from torch_geometric.typing import FeatureTensorType
+from torch_geometric.utils.mixin import CastMixin
+
+
+@dataclass
+class TensorAttr(CastMixin):
+    r"""Defines the attributes of a :obj:`FeatureStore` tensor."""
+
+    # The node indices the rows of the tensor correspond to
+    index: Optional[FeatureTensorType] = None
+
+    # The type of the feature tensor (may be used if there are multiple
+    # different feature tensors for the same node index)
+    tensor_type: Optional[str] = None
+
+    # The type of the nodes that the tensor corresponds to (may be used for
+    # hetereogeneous graphs)
+    node_type: Optional[str] = None
+
+    # The type of the graph that the nodes correspond to (may be used if a
+    # feature store supports multiple graphs)
+    graph_type: Optional[str] = None
+
+
+class AttrView:
+    r"""A view of a :obj:`FeatureStore` that is obtained from an incomplete
+    specification of attributes. This view stores a reference to the
+    originating feature store as well as a :obj:`TensorAttr` object that
+    represents the view's (incomplete) state.
+
+    As a result, store[TensorAttr(...)].tensor_type[idx] allows for indexing
+    into the store.
+    """
+    _store: 'FeatureStore'
+    attr: TensorAttr
+
+    def __init__(self, store, attr):
+        self._store = store
+        self.attr = attr
+
+    def __getattr__(self, tensor_type):
+        r"""Supports attr_view.attr"""
+        self.attr.tensor_type = tensor_type
+        return self
```

Comment:
not 100% sure where did  we use this.

Reply:
This is to support advanced indexing, see [these lines](https://github.com/pyg-team/pytorch_geometric/pull/4534/files#diff-f591068f1a527a7fde0baa34426082c764ddba485da2aa2346f0cf3ca29981dcR84-R86).

Reply:
I feel advance indexing is cool and let `FeatureStore` inherit from `MutableMapping` is also cool but learning and understand curve is higher.
The tricky part is this indexing only support `tensor_type` but what if we want to do things like `.get(node_type=..., tensor_type=...)` or `.get(node_type=...)`. I think it is a little hard to follow the code if we just read the api.
Similar thing apply to `FeatureStore` where `store[...]` can return either `tensors` or `AttrView`. It will make it hard to read code and also do type check.

Reply:
Thanks for the comments! Let me clarify a couple points here. 
* The current design of the feature store assumes that all get calls require a `TensorAttr.index`.
* A `get_tensor` call will either return a Tensor (if found, based on `TensorAttr`), `None` if the tensor is not found (because the attributes were not fully specified, or because the attributes do not match a stored value), or raise an Exception if `index` is `None`. I think this behavior makes sense and aligns with what a user would expect from such a function.
* An indexing call will return a Tensor if the attributes are fully specified (`index` is not `None`). It will return an `AttrView` if the attributes are not fully specified (that is, if `index` is `None`), which allows for further specification of the tensor type and the index via `__getattr__` and `__getitem__`. If the tensor type is specified in `TensorAttr` at the initial indexing call, one can simply do `store[{'node_type': node_type, 'tensor_type': tensor_type}][index]` and obtain the desired values. Imo, it seems a bit confusing to support `store[{'tensor_type': tensor_type}].node_type[index]` as this pattern seems out-of-order (first specifying the tensor type, then a node type): a user must first specify general attributes, then a node type, then an index, which is what we currently support. However, if this is desired, it would be quite easy to implement it.

Wdyt?

Reply:
I think a `AttrView` is a way of select features from store.
```
a = AttrView(store, {'node_type': x, 'graph_type': y})
b = a.feat
c = a.pkey
# b and c should be different views.
```
In another word, I would prefer not expose getattr but explicitly specify views like this
```
a = AttrView(store, {'node_type': x, 'graph_type': y})
b = AttrView(store, {'node_type': x, 'graph_type': y, 'tensor_type': 'feat'})
c = AttrView(store, {'node_type': x, 'graph_type': y, 'tensor_type': 'pkey'})
```
More over, for the same node_type and graph_type, how about support tensor_type as a list? So that if we decide to implement feature store on a different host, when we send a view from a source host to a destination host, we only pass index once for multiple tensor types.

For feature store, I'd prefer have explicit functions for views and get_tensors, i.e keep get_tensor as it is but also define a function
```
def view(ta: TensorAttr) -> AttrView:
```
It has two benefit, each function will have explicit return type. And in addition to it, the function name also reflect the function workload. Create an `AttrView` by `view` function should be lighter than create a `Tensor` by `get_tensor` function. We don't mix them together by `__getitem__` function.

Reply:
Closing in favor of the revised design.